contracts/treasury/contracts/src/ZrxTreasury.sol

```diff
@@ -0,0 +1,433 @@
+    ///      given proposal ID has passed. Returns false if
+    ///      there is no proposal associated with the given ID.
+    ///      For a proposal to pass, it must receive more votes
+    ///      for than against, and the number of votes for must
+    ///      meet the `quorumThreshold`.
+    /// @param proposalId The ID of the proposal.
+    /// @return hasPassed Whether the proposal has passed.
+    function hasProposalPassed(uint256 proposalId)
+        external
+        override
+        view
+        returns (bool hasPassed)
+    {
+        if (proposalId >= proposalCount()) {
+            return false;
+        }
+        Proposal memory proposal = proposals[proposalId];
+        return _hasProposalPassed(proposal);
+    }
+
+    /// @dev Creates a proposal to send ZRX from this treasury on the
+    ///      the given actions. Must have at least `proposalThreshold`
+    ///      of voting power to call this function. See `getVotingPower`
+    ///      for how voting power is computed. If a proposal is successfully
+    ///      created, voting starts at the epoch after next (currentEpoch + 2).
+    ///      If the vote passes, the proposal is executable during the
+    ///      `executionEpoch`. See `hasProposalPassed` for the passing criteria.
+    /// @param actions The proposed ZRX actions. A action specifies an
+    ///        amount of ZRX to award to a specified recipient address.
+    /// @param executionEpoch The epoch during which the proposal is to
+    ///        be executed if it passes. Must be at least two epochs
+    ///        from the current epoch.
+    /// @param description A text description for the proposal.
+    /// @param operatedPoolIds The pools operated by `account`. The
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    /// @return proposalId The ID of the newly created proposal.
+    function propose(
+        ProposedAction[] memory actions,
+        uint256 executionEpoch,
+        string memory description,
+        bytes32[] memory operatedPoolIds
+    )
+        public
+        override
+        returns (uint256 proposalId)
+    {
+        require(
+            getVotingPower(msg.sender, operatedPoolIds) >= proposalThreshold,
+            "propose/INSUFFICIENT_VOTING_POWER"
+        );
+        require(
+            actions.length > 0,
+            "propose/NO_ACTIONS_PROPOSED"
+        );
+        uint256 currentEpoch = stakingProxy.currentEpoch();
+        require(
+            executionEpoch >= currentEpoch + 2,
+            "propose/INVALID_EXECUTION_EPOCH"
+        );
+
+        proposalId = proposalCount();
+        Proposal storage newProposal = proposals.push();
+
+        for (uint256 i = 0; i != actions.length; i++) {
+            newProposal.actions.push(actions[i]);
+        }
+        newProposal.executionEpoch = executionEpoch;
+        newProposal.voteEpoch = currentEpoch + 2;
```

Comment:
what's the reason why voting starts in two epochs?

Reply:
we want to give ppl enough time  delegate their ZRX before the vote. otherwise someone can propose something at the very end of an epoch and everyone who wasn't already delegated wouldn't be eligible to vote